Allen n/tf tts

allen-n opened the pull request:
It's done! EPUBs can be converted to speech completely offline (and free!).

Looking forward to your comments on improvements I can make to make the code cleaner/run better.

Brian-Lam commented on README.md:
```diff
@@ -25,9 +42,27 @@ $ virtualenv venv
+### Offline model downloads
+**Option 1: Download pre-zipped models**
+
+Download [the zip file](https://drive.google.com/file/d/1fcc8Rdl9WemkwFoiLg4p22vilwd5h8KV/view?usp=sharing) and extract to this directory.
```
Can you clarify this? Do you mean the root directory of the repo?

allen-n replied:
Yes you're right it is in fact repo root- added that.

Brian-Lam commented on convert.py:
```diff
@@ -29,10 +30,20 @@
     default="", 
     help="The filepath of the .epub file to convert"
 )
+parser.add_argument("--offline", 
```
I know this is not the most convenient change to make, but I'm not sure if "offline" is the right term here, it seems a little misleading

Maybe --local ?

Offline may give the impression that this functionality is meant to be run when the computer is offline (e.g. a fallback when there is no internet connection), as opposed to the fact that it runs Speech Synthesis locally

allen-n replied:
No I think that's a good point I struggled with coming up with the right word here, and it was a pretty easy find & replace.

Brian-Lam commented on convert.py:
```diff
@@ -29,10 +30,20 @@
 # The filepath of the eBook to convert.
 EPUB_FILEPATH = args.file
+ENABLE_OFFLINE = args.offline
+if not ENABLE_OFFLINE:
```
Love the dynamic imports! Personally I'd suggest:

if ENABLE_OFFLINE:
    from tts_utils import TTS
    import re
else:
   from google.cloud import texttospeech

to avoid the double negative (just fewer mental gymnastics, since we have an ```if not``` and an else)

allen-n replied:
ack

Brian-Lam commented on convert.py:
```diff
@@ -104,6 +115,28 @@ def GenerateAudioContentForText(text):
             break
     
 
+def eupub_to_chapters():
```
Some function documentation above this would be great :)

allen-n replied:
ack

Brian-Lam commented on convert.py:
```diff
@@ -104,6 +115,28 @@ def GenerateAudioContentForText(text):
+    full_text = ''
+    for text in book.get_items_of_type(ebooklib.ITEM_DOCUMENT):
+        html_content = text.get_content().decode("utf-8")
+        full_text += html_content
```
This is a famously and deceptive expensive operation. Since strings are immutable in Python, every single time you "append" content to the string, it doesn't actually modify the existing string - it actually performs a deep copy, and returns a reference to the new variable with all the data.

For more reference:
https://stackoverflow.com/questions/53236784/efficient-string-concatenation-in-python-3

So a deep copy of all the data that has already been read is being performed in every iteration of the loop. Since we don't have a guaranteed ceiling on the size of the input file or its structure (number of ITEM_DOCUMENTS), this could actually get pretty expensive. 

Solution would be to do something like:

```
chapter_texts = []
for text in book.get_items_of_type(ebooklib.ITEM_DOCUMENT):
    chapter_texts.append(text.get_content().decode("utf-8"))
full_text = "".join(full_text)
```

allen-n replied:
Good catch! This is really good to know generally - ty (fixed)

Brian-Lam commented on convert.py:
```diff
@@ -104,6 +115,28 @@ def GenerateAudioContentForText(text):
+    if OUTPUT_EPUB_TEXT_TO_TEST_FILE:
+        f = open("chapter_test.txt", "w")
+        for i, chapter in enumerate(chapter_text):
+            f.write("!!Chapter {}:\n{}".format(i+1,chapter))
```
A quick comment would be appropriate here to explain this formatting :)

allen-n replied:
Ack - #allen's weird debug strings

Brian-Lam commented on convert.py:
```diff
@@ -202,6 +254,7 @@ def convert_text_chunks_to_speech(ebook_chunks):
     chunk_mp3_filepaths = []
     for chunk in ebook_chunks:
         if (chunks_converted < MAX_CHUNK_COUNT):
+            chunk_mp3_filepath = None
```
Is this necessary?

allen-n replied:
It is not - not sure how it snuck in there!

Brian-Lam commented on convert.py:
```diff
@@ -210,14 +263,67 @@ def convert_text_chunks_to_speech(ebook_chunks):
+    path = os.path.abspath(path)
+    print("Joining audio for {}".format(out_filename))
+    for filename in tqdm(sorted(os.listdir(path))):
+        # print("filename: {}, abs: {}, path: {}".format(filename, os.path.abspath(filename), path))
```
Can we remove the commented out code? :)

allen-n replied:
Yes - bad habit of mine, I'l; clean out the other files too

Brian-Lam commented on convert.py:
```diff
@@ -210,14 +263,67 @@ def convert_text_chunks_to_speech(ebook_chunks):
+    path = os.path.abspath('./audio')
+    
+    for i, chapter in enumerate(ebook_chunks):
+        # s = "Chapter_{}_audio".format(i)
```
Ditto

allen-n replied:
ack

Brian-Lam commented on convert.py:
```diff
@@ -210,14 +263,67 @@ def convert_text_chunks_to_speech(ebook_chunks):
     print("All chunks converted.")
     return chunk_mp3_filepaths
 
+def merge_wavs_to_mp3(out_filename, path=".", ms_delay=100):
```
Function documentation would be great :)

allen-n replied:
ack

Brian-Lam commented on convert.py:
```diff
@@ -210,14 +263,67 @@ def convert_text_chunks_to_speech(ebook_chunks):
+        if os.path.splitext(filename)[1] == ".wav":
+            os.remove(os.path.join(path, filename))
+
+def offline_text_chunks_to_speech(ebook_chunks):
```
Function documentation would be great :)

allen-n replied:
ack

Brian-Lam commented on convert.py:
```diff
@@ -210,14 +263,67 @@ def convert_text_chunks_to_speech(ebook_chunks):
-        merge_chunk_mp3s(chunk_mp3_filepaths)
-        delete_chunk_mp3s(chunk_mp3_filepaths)
\ No newline at end of file
+    # tts = TTS(tts='fastspeech2', generator='multiband_melgan2_generator')
```
Can we remove the commented out code?

Brian-Lam reviewed (changes requested):
Hey - thanks a bunch for this, this is awesome and I'm genuinely super impressed! Left a few comments.... Lmk what you think, feel free to ask if you have any questions!

allen-n commented:
Thanks! I'm not sure what the proper flow is for pushing up new changes and resolving the old ones, but I _think_ I addressed all of your comments. Take a peek whenever convenient (or better yet - convert something on your device locally :) - thanks again for the feedback here.

Brian-Lam commented on convert.py:
```diff
@@ -114,29 +115,38 @@ def GenerateAudioContentForText(text):
     
-
 def eupub_to_chapters():
+    """Parse a .epub file and return an array of strings
```
Nit: newline after """

Brian-Lam commented on convert.py:
```diff
@@ -114,29 +115,38 @@ def GenerateAudioContentForText(text):
+        a chapter of the .epub file, i.e. ["chapter 1 text", "chapter 2 text", ...]
+    """    
     path = os.path.abspath(EPUB_FILEPATH)
     print(path)
```
Recommend either taking this out or making a more informative print statement, e.g.:

print("Reading EPUB from filepath: " + path)

Brian-Lam commented on convert.py:
```diff
@@ -203,13 +213,17 @@ def full_text_to_chunks(full_text):
-@param num_senrdr    tences: the number of sentences to add to each chunk
-"""
 def chapters_to_chunks(chapters, num_sentences=4):
+    """Breaks down a book into chunks of ~MAX_CHUNK_LENGTH characters each, not
```
Nit: newline after """
